src/sql/src/plan/query.rs

```diff
@@ -2968,24 +2966,32 @@ fn plan_case<'a>(
 fn plan_literal<'a>(l: &'a Value) -> Result<CoercibleScalarExpr, anyhow::Error> {
     let (datum, scalar_type) = match l {
         Value::Number(s) => {
-            let d: Decimal = s.parse()?;
-            if d.scale() == 0 {
-                let significand = d.significand();
-                if let Ok(n) = significand.try_into() {
-                    (Datum::Int32(n), ScalarType::Int32)
-                } else if let Ok(n) = significand.try_into() {
-                    (Datum::Int64(n), ScalarType::Int64)
-                } else {
-                    (
-                        Datum::from(significand),
-                        ScalarType::Decimal(MAX_DECIMAL_PRECISION, d.scale()),
-                    )
+            match strconv::parse_numeric(s.as_str()) {
+                Ok(d) => {
+                    if !s.contains(&['E', '.'][..]) {
```

Comment:
I don't believe so because `9e0` should not be convertible to an int, but there's nothing intrinsic to the apd value to indicate that was the string that spawned it, even though the value cleanly converts to an int. Similarly, `1.0` should not be convertible to an int, but determining that it's ineligible requires looking at the value's exponent--but I was already looking at the string for `E`, so figured it was nominally cleaner to just check for periods in the same spot.